runtime/gc_vlhgc/WriteOnceCompactor.cpp

```diff
@@ -1382,42 +1382,54 @@ MM_WriteOnceCompactor::fixupPointerArrayObject(MM_EnvironmentVLHGC* env, J9Objec
 	_extensions->classLoaderRememberedSet->rememberInstance(env, objectPtr);
 
 	/* arraylet leaves are walked separately in fixupArrayletLeafRegionContentsAndObjectLists(), to increase parallelism. Just walk the spine */
-	GC_ArrayletObjectModel::ArrayLayout layout = _extensions->indexableObjectModel.getArrayLayout((J9IndexableObject*)objectPtr);
-		
+	GC_ArrayObjectModel *indexableObjectModel = &_extensions->indexableObjectModel;
+	GC_ArrayletObjectModel::ArrayLayout layout = indexableObjectModel->getArrayLayout((J9IndexableObject *)objectPtr);
+
 	if (GC_ArrayletObjectModel::InlineContiguous == layout) {
-		UDATA elementsToWalk = _extensions->indexableObjectModel.getSizeInElements((J9IndexableObject*)objectPtr);
-		GC_PointerArrayIterator it(_javaVM, objectPtr);
-		UDATA previous = 0;
-		for (UDATA i = 0; i < elementsToWalk; i++) {
-			GC_SlotObject *slotObject = it.nextSlot();
-			Assert_MM_true(NULL != slotObject);
-			J9Object* pointer = slotObject->readReferenceFromSlot();
-			if (NULL != pointer) {
-				J9Object* forwardedPtr = getForwardWrapper(env, pointer, cache);
-				slotObject->writeReferenceToSlot(forwardedPtr);
-				_interRegionRememberedSet->rememberReferenceForCompact(env, objectPtr, forwardedPtr);
+		/* For offheap we need a special check for the case of a partially offheap allocated array that caused the current GC
+		 * (its dataAddr field will still be NULL), with offheap heap we will fixup camouflaged discontiguous arrays) - DM, like default
+		 * balanced, wants to fixup only truly contiguous arrays
+		 */
+		if (indexableObjectModel->isDataAdjacentToHeader((J9IndexableObject *)objectPtr)
+#if defined(J9VM_ENV_DATA64)
+			|| (indexableObjectModel->isVirtualLargeObjectHeapEnabled()
+				&& (NULL != indexableObjectModel->getDataAddrForContiguous((J9IndexableObject *)objectPtr)))
+#endif /* defined(J9VM_ENV_DATA64) */
+		 ) {
+			uintptr_t elementsToWalk = indexableObjectModel->getSizeInElements((J9IndexableObject *)objectPtr);
+			GC_PointerArrayIterator it(_javaVM, objectPtr);
+			uintptr_t previous = 0;
+			for (uintptr_t i = 0; i < elementsToWalk; i++) {
+				GC_SlotObject *slotObject = it.nextSlot();
+				Assert_MM_true(NULL != slotObject);
+				J9Object *pointer = slotObject->readReferenceFromSlot();
+				if (NULL != pointer) {
+					J9Object *forwardedPtr = getForwardWrapper(env, pointer, cache);
+					slotObject->writeReferenceToSlot(forwardedPtr);
```

Comment:
Since we are fixing up any ref array here (adjacent or not), we don't need this pass that is specific for hybrid arraylets: fixupArrayletLeafRegionSpinePointers

It's not harm to do it, but it's an unnecessary region iteration (and unnecessary modification of that method)

Similarly, a part of fixupArrayletLeafRegionContentsAndObjectLists will be dead that fixes up leafs, but since the region iteration also piggybacks some other work, we cannot say that pass is completely dead. But parhps we can assert offheap is disabled in the the block that fixes up leafs.

Reply:
we should also add split array logic for parallel scanning

Reply:
> Similarly, a part of fixupArrayletLeafRegionContentsAndObjectLists will be dead that fixes up leafs, but since the region iteration also piggybacks some other work, we cannot say that pass is completely dead. But parhps we can assert offheap is disabled in the the block that fixes up leafs.

It' likely we indeed enter that block/loop where we fixup leaf references, and we should skip it for offheap enabled. It's a correctness problem, since we might be finding garbage references to scan. Additionally, because ther are supposed to be left decommited till the array dies, we are recommiting them too early.